src/dataregistry/registrar_util.py

```diff
@@ -1,30 +1,44 @@
 import os
+import re
+from sqlalchemy import MetaData, Table, Column, text, select
+
 from dataregistry.db_basic import ownertypeenum
 
-__all__ = ['make_version_string', 'parse_version_string']
+__all__ = ['make_version_string', 'parse_version_string', 'calculate_special',
+           'form_dataset_path', 'get_directory_info']
 VERSION_SEPARATOR = '.'
+_nonneg_int_re = "0|[1-9][0-9]*"
+
 def make_version_string(major, minor=0, patch=0, suffix=None):
     version = VERSION_SEPARATOR.join([major, minor, patch])
     if suffix:
         version = VERSION_SEPARATOR.join([version, suffix])
     return version
 
-def parse_version_string(version):
+def parse_version_string(version, with_suffix=False):
     '''
-    Return dict with keys major, minor, patch and (if present) suffix
+    Return dict with keys major, minor, patch and (if present) suffix.
+    with_suffix == False means version string *must not* include suffix
+    with_suffix == True means it *may* have a suffix
+
+    Returns a dict with keys 'major', 'minor', 'patch' and optionally 'suffix'
     '''
     # Perhaps better to do with regular expressions.  Or at least verify
     # that major, minor, patch are integers
-    cmp = version.split(_SEPARATOR, maxsplit=3)
-    d = {'major' : cmp[0]}
-    if len(cmp) > 1:
-        d['minor'] = cmp[1]
+    cmp = version.split(VERSION_SEPARATOR)
+    if not with_suffix:
+        if len(cmp) != 3:
+            raise ValueError('Version string must have 3 components')
     else:
-        d['minor'] = 0
-    if len(cmp) > 2:
-        d['patch'] = cmp[2]
-    else:
-        d['patch'] = 0
+        if len(cmp) < 3 or len(cmp) > 4:
```

Comment:
How do i construct a version with suffix in the case of `len(cmp) == 3`, wont it clash with the regex check?

Reply:
The regex check is only done on cmp[0:3] (does not include cmp[3])

Reply:
ok, I had in my head version strings with suffixes were along the lines of "1.2.3-bob" over "1.2.3.bob".

I guess the user will never be using this function (and its not used with suffixes by us yet either)

Reply:
Right: this function is really not meant for users. I will change the name to `_parse_version_string`